cmd/podman/spec.go

```diff
@@ -598,3 +607,43 @@ func makeThrottleArray(throttleInput []string) ([]spec.LinuxThrottleDevice, erro
 	}
 	return ltds, nil
 }
+
+// CreatePortBindings iterates ports mappings and exposed ports into a format CNI understands
+func (c *createConfig) CreatePortBindings() ([]ocicni.PortMapping, error) {
+	var portBindings []ocicni.PortMapping
+	for containerPb, hostPb := range c.PortBindings {
+		var PM ocicni.PortMapping
+		PM.ContainerPort = int32(containerPb.Int())
+		for _, i := range hostPb {
+			var hostPort int
+			var err error
+			PM.HostIP = i.HostIP
+			if i.HostPort == "" {
+				hostPort = containerPb.Int()
+			} else {
+				hostPort, err = strconv.Atoi(i.HostPort)
+				if err != nil {
+					return portBindings, errors.Wrapf(err, "unable to convert host port to integer")
+				}
+			}
+
+			PM.HostPort = int32(hostPort)
+			// CNI requires us to make both udp and tcp structs
+			PM.Protocol = "udp"
+			portBindings = append(portBindings, PM)
+			PM.Protocol = "tcp"
```

Comment:
Does this work? I don't know if append is guaranteed to force a copy on structs

Reply:
yes it works